spec/index.bs

```diff
@@ -6678,10 +6721,66 @@ interface mixin GPUProgrammablePassEncoder {
             {{GPUBufferBindingLayout/minBindingSize}}, the binding ranges are large enough for
             the shader's minimum binding size requirements.
         </div>
+    1. [$Encoder bind groups alias a writable resource$](|encoder|, |pipeline|) must be `false`.
+
+        Issue(gpuweb/gpuweb#1842): Determine what happens when an application violates this.
+        Is it a validation error, one of multiple possible behaviors, or do we just remove this
+        restriction entirely and allow writable bindings to alias with undefined results?
 
     Otherwise return `true`.
 </div>
 
+<div algorithm>
+    <dfn abstract-op>Encoder bind groups alias a writable resource</dfn>(|encoder|, |pipeline|)
+    if any writable buffer binding range overlaps with any other binding range of the same buffer,
+    or any writable texture binding overlaps in [=texture subresources=] with any other texture binding
+    (which may use the same or a different {{GPUTextureView}} object).
+
+    **Arguments:**
+    <dl>
+        : {{GPUProgrammablePassEncoder}} |encoder|
```

Comment:
Sorry to bring this up late. We need to figure out what the scope of aliasing analysis is.
There are several choices:
  1. scope is the synchronization scope, e.g. the whole render pass.
  2. scope is the draw call (i.e. the pipeline with its active bindings).
  3. scope is a single entry point and its bindings.

My understanding is that only (3) makes sense, because that's what the program in Metal is build for, and we are only talking about UB from the point of view of the shader compiler.

If we are talking about UB from the point of data races, than this has been agreed previously to be allowed, and I don't think we are going to change this.

Reply:
> My understanding is that only (3) makes sense, because that's what the program in Metal is build for, and we are only talking about UB from the point of view of the shader compiler.

This is a good point and brings some clarity to what we're doing here. I think I agree, so I can add back the foreach-stage loop.

However what I would like to understand better is what optimization the shader compiler can possibly make that results in problems when a resource is aliased in one entry point but is OK when a resource is aliased in any other way (between entry points or between draw calls).

Reply:
> However what I would like to understand better is what optimization the shader compiler can possibly make that results in problems when a resource is aliased in one entry point but is OK when a resource is aliased in any other way (between entry points or between draw calls).

I don't think there's any difference (and the race from another draw if anything will be harder for the compiler to reason about), so that's why I don't think we should have this validation at all.

Reply:
It makes a ton of difference from the point of view of TBAA. Consider this code:
```rust
let foo = buffer1.x;
buffer2.x = 2;
if (foo) { discard; }
```

If the compiler knows `buffer1` and `buffer2` aren't aliased, it also knows that there is no operations affecting `foo` result between the place it's declared and used. It can move the actual load into the `if` condition, for example. I.e. it clearly affects the ability of compilers to optimize.

Reply:
That's right, and it is a bit of a developer footgun. However if you're using read-write storage buffers in FS you already have a footgun with the data races from other shaders. So maybe it's ok and people doing fancy algorithms in storage buffers will have to be cognizant of the memory models.

Reply:
Right. In principle, I agree with this. But one of the major points in the atomics discussion was that the UB for compiler is strictly separate from the UB of the run-time. We keep the latter, but we can't afford the former - that's the status quo today.